apps/src/templates/manageStudents/ManageStudentsLoginInfo.jsx

```diff
@@ -15,13 +16,47 @@ class ManageStudentsLoginInfo extends Component {
     sectionId: PropTypes.number,
     sectionCode: PropTypes.string,
     loginType: PropTypes.string,
-    // The prefix for the code studio url in the current environment,
-    // e.g. '//studio.code.org' or '//localhost-studio.code.org:3000'.
-    studioUrlPrefix: PropTypes.string,
+  };
+
+  prependUrlOrigin = (url) => {
+    if (url.startsWith('//')) {
+      return `http:${url}`;
+    } else {
+      return window.location.origin + url;
```

Comment:
I have some gripes with this method. It appears to give a different protocol for `joinSectionUrl` links depending on whether it's running on dashboard or pegasus. It's also adding a protocol in one case and an origin in another, contradictory to the name and without any comment explaining why this would be.

If `https: is ok, you could remove these 2 lines and add an optional param to `pegasus` and `studio` e.g. `function pegasus(url, includeOrigin = false)` to give you the right url in one go.

If keeping the `http:` on these urls is a requirement, then I'd suggest finding another way to separate the code more clearly that's trying to generate urls with `http:` from those that are not.

Reply:
Another approach is to lean on the browser API here; the [Location](https://developer.mozilla.org/en-US/docs/Web/API/Location) API is fairly powerful for handling URLs, and headless anchor elements implement it:

![image](https://user-images.githubusercontent.com/1615761/51877025-6eb7ac00-231f-11e9-8b52-11f3797b2fff.png)

Reply:
yeah, i see what you mean. i don't think the `pegasus`/`studio` helpers will work in this case because they rely on `window.pegasus` and `window.dashboard` being present, and those objects (pegasus & dashboard) are only defined in the _opposite_ environment (i.e., `window.pegasus` is available in dashboard, and `window.dashboard` is available in pegasus). so if i call the `studio` method from dashboard, i don't have access to `window.pegasus.STUDIO_URL`, which i would need to always include the origin.

the method i wrote (`prependUrlOrigin`) was a way to get around this limitation to format the URL for the user and display it in a readable format here (that's why it's only used by the `joinSectionUrl` and `sectionSignInUrl` methods:

![screen shot 2019-01-28 at 5 13 40 pm](https://user-images.githubusercontent.com/9812299/51877275-788ddf00-2320-11e9-8623-af6aff13e92e.png)

we always want to display the absolute path here, so this helper method was a way to achieve that. i definitely agree that the naming of this method is incorrect, since sometimes it prepends `http:` and sometimes it prepends the origin, but i thought maybe some context around the issue i'm solving for might help :)

Reply:
> yeah, i see what you mean. i don't think the pegasus/studio helpers will work in this case because they rely on window.pegasus and window.dashboard being present, and those objects (pegasus & dashboard) are only defined in the opposite environment (i.e., window.pegasus is available in dashboard, and window.dashboard is available in pegasus). so if i call the studio method from dashboard, i don't have access to window.pegasus.STUDIO_URL, which i would need to always include the origin.

You can deal with this the same way you do it in prependUrlOrigin -- by looking at the window.location.origin to get the origin of the current environment. 

What I'm pushing for here is to refactor the existing functionality to have methods with abstractions that are easy to reason about. I'm pushing for something like this: 
```
// returns a pegasus url that is valid for the current environment
function pegasus(path)

// returns a fully qualified pegasus url
function fullPegasusUrl(path)
```

not this:
```js
// returns a pegasus url that is valid for the current environment
function pegasus(path)

// given a url that was either a valid pegasus or dashboard url
// that is valid in the current environment, returns a fully qualified url.
function prependUrlOrigin(...)
```

I'm making a point of this because I believe this is an instance of the common pattern of code being too DRY. We should prefer having abstractions that are easy to reason about over adhering to the DRY principle:

* https://fluent.software/abstraction-vs-duplication/
* https://www.sandimetz.com/blog/2016/1/20/the-wrong-abstraction

Sorry to make a stink. If we disagree in principle then I'm open to discussing, but I think this could be a helpful principle for all of us to keep in mind to make our code as readable as possible.

Reply:
oh, i totally see what you're saying now. good call, i'll go that route. and don't be sorry! i appreciate you pushing back on this